harness/determined/core/_context.py

```diff
@@ -318,3 +325,21 @@ def init(
         experimental=experimental,
         _tensorboard_manager=tensorboard_manager,
     )
+
+
+def run_prepare(
```

Comment:
Makes sense, also switched the test to use the distrusted example since it checked everything and more

This also applies to context_v1 right? I implemented the same ```_run_on_rank_0_and_broadcast``` logic for the v1 context.

As a side note, I wasn't sure how to import from core_v2 since it was giving me import errors and I am not familiar enough with the directory structure to find where to move the ```_run_on_rank_0_and_broadcast```. Do you have any advice on what to do about importing it or if just duplicating the logic in that one spot is okay?

Reply:
thinking out loud here: if a user does not pass `checkpoint_storage` parameter, and it's running in managed mode for standard experiments (i.e. context v1 and managed mode in v2). in this case, the storage is always known by master in advance (from the submitted experiment config), and so the storage backend can be created inside `CreateExperiment`, and then the storage id is passed through an env variable to the context's `TrialInfo` class. this'll avoid one extra API call. it's an optimization but I am not sure it's worth it.

>As a side note, I wasn't sure how to import from core_v2 since it was giving me import errors and I am not familiar enough with the directory structure to find where to move the _run_on_rank_0_and_broadcast. Do you have any advice on what to do about importing it or if just duplicating the logic in that one spot is okay?

can you put it into `determined/core/_distributed.py`? it's tightly integrated with the distributed context after all. imports today are mostly from v1 to v2 so a reverse may likely cause circular imports problem.

Reply:
moved into ```determined/core/_distributed.py```

Reply:
on the optimization question I'm not sure either

I thought part of the idea is to bundle all these operations into a single API call so even if we do optimize out the storageID call for the common case then we may still have to make the API call. I'm not exactly sure what these other API calls would do and how soon we would need to add them though.

I'm open to adding the optimization if you think it is worth it though

Reply:
Agreed on bundling. I am leaning towards "not worth it". I'll make a note of it in a ticket: https://hpe-aiatscale.atlassian.net/browse/DET-10096 so it's not immediately forgotten.